File: example04retrofit/src/main/java/foo/bar/example/asafretrofit/ui/fruit/FruitView.java
package foo.bar.example.asafretrofit.ui.fruit;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

import butterknife.BindView;
import butterknife.ButterKnife;
import co.early.fore.core.callbacks.FailureCallbackWithPayload;
import co.early.fore.core.callbacks.SuccessCallback;
import co.early.fore.core.observer.Observer;
import co.early.fore.core.ui.SyncableView;
import foo.bar.example.asafretrofit.CustomApp;
import foo.bar.example.asafretrofit.R;
import foo.bar.example.asafretrofit.feature.fruit.FruitFetcher;
import foo.bar.example.asafretrofit.message.UserMessage;
import foo.bar.example.asafretrofit.ui.widgets.AnimatedTastyRatingBar;

/**
 *
 */
public class FruitView extends ScrollView implements SyncableView{

    //models that we need to sync with
    private FruitFetcher fruitFetcher;


    //UI elements that we care about
    @BindView(R.id.fruit_fetchsuccess_btn)
    public Button fruitRefreshSuccess;

    @BindView(R.id.fruit_fetchfailbasic_btn)
    public Button fruitRefreshFailBasic;

    @BindView(R.id.fruit_fetchfailadvanced_btn)
    public Button fruitRefreshFailAdvanced;

    @BindView(R.id.fruit_name_textview)
    public TextView fruitName;

    @BindView(R.id.fruit_citrus_img)
    public ImageView isCitrus;

    @BindView(R.id.fruit_tastyrating_tastybar)
    public AnimatedTastyRatingBar tasteValuePercentView;

    @BindView(R.id.fruit_tastyrating_textview)
    public TextView tasteScore;

    @BindView(R.id.fruit_busy_progbar)
    public ProgressBar busy;

    @BindView(R.id.fruit_detailcontainer_linearlayout)
    public View fruitDetailContainer;


    //single observer reference
    Observer observer = this::syncView;



    //just because we re-use these in 3 different button clicks
    //in this example we define them here
    private SuccessCallback successCallback = new SuccessCallback() {
        @Override
        public void success() {
            Toast.makeText(getContext(), "Success - you can use this trigger to " +
                    "perform a one off action like starting a new activity or " +
                    "something", Toast.LENGTH_SHORT).show();
        }
    };
    private FailureCallbackWithPayload<UserMessage> failureCallback = new FailureCallbackWithPayload<UserMessage>() {
        @Override
        public void fail(UserMessage userMessage) {
            Toast.makeText(getContext(), "Fail - maybe tell the user to try again, message:" + userMessage.getString(),
                    Toast.LENGTH_SHORT).show();
        }
    };





    public FruitView(Context context) {
        super(context);
    }

    public FruitView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public FruitView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }



    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        ButterKnife.bind(this, this);

        getModelReferences();

        setupButtonClickListeners();
    }

    private void getModelReferences(){
        fruitFetcher = CustomApp.get(FruitFetcher.class);
    }

    private void setupButtonClickListeners() {

        fruitRefreshSuccess.setOnClickListener(v -> fruitFetcher.fetchFruits(successCallback, failureCallback));

        fruitRefreshFailBasic.setOnClickListener(v -> fruitFetcher.fetchFruitsButFailBasic(successCallback, failureCallback));

        fruitRefreshFailAdvanced.setOnClickListener(v -> fruitFetcher.fetchFruitsButFailAdvanced(successCallback, failureCallback));

    }


    //data binding stuff below

    public void syncView(){
        fruitRefreshSuccess.setEnabled(!fruitFetcher.isBusy());
        fruitRefreshFailBasic.setEnabled(!fruitFetcher.isBusy());
        fruitRefreshFailAdvanced.setEnabled(!fruitFetcher.isBusy());
        fruitName.setText(fruitFetcher.getCurrentFruit().name);
        isCitrus.setImageResource(fruitFetcher.getCurrentFruit().isCitrus ? R.drawable.lemon_positive : R.drawable.lemon_negative);
        tasteValuePercentView.setTastyPercent(fruitFetcher.getCurrentFruit().tastyPercentScore);
        tasteScore.setText(String.format(getContext().getString(R.string.fruit_percent), String.valueOf(fruitFetcher.getCurrentFruit().tastyPercentScore)));
        busy.setVisibility(fruitFetcher.isBusy() ?  VISIBLE : INVISIBLE);
        fruitDetailContainer.setVisibility(fruitFetcher.isBusy() ? GONE : VISIBLE);
    }


    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        fruitFetcher.addObserver(observer);
        syncView(); //  <- don't forget this
    }


    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        fruitFetcher.removeObserver(observer);
    }
}

File: example04retrofitcoroutine/src/main/java/foo/bar/example/foreretrofitcoroutine/ui/fruit/FruitView.kt
package foo.bar.example.foreretrofitcoroutine.ui.fruit

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.ScrollView
import android.widget.Toast
import co.early.fore.core.callbacks.FailureWithPayload
import co.early.fore.core.callbacks.Success
import foo.bar.example.foreretrofit.R
import foo.bar.example.foreretrofitcoroutine.OG
import foo.bar.example.foreretrofitcoroutine.feature.fruit.FruitFetcher
import foo.bar.example.foreretrofitcoroutine.message.UserMessage
import kotlinx.android.synthetic.main.fragment_fruit.view.fruit_busy_progbar
import kotlinx.android.synthetic.main.fragment_fruit.view.fruit_citrus_img
import kotlinx.android.synthetic.main.fragment_fruit.view.fruit_detailcontainer_linearlayout
import kotlinx.android.synthetic.main.fragment_fruit.view.fruit_fetchchainsuccess_btn
import kotlinx.android.synthetic.main.fragment_fruit.view.fruit_fetchfailadvanced_btn
import kotlinx.android.synthetic.main.fragment_fruit.view.fruit_fetchfailbasic_btn
import kotlinx.android.synthetic.main.fragment_fruit.view.fruit_fetchsuccess_btn
import kotlinx.android.synthetic.main.fragment_fruit.view.fruit_name_textview
import kotlinx.android.synthetic.main.fragment_fruit.view.fruit_tastyrating_tastybar
import kotlinx.android.synthetic.main.fragment_fruit.view.fruit_tastyrating_textview

/**
 *
 */
class FruitView @JvmOverloads constructor(
        context: Context?,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : ScrollView(context, attrs, defStyleAttr) {

    //models that we need to sync with
    private lateinit var fruitFetcher: FruitFetcher


    //single observer reference
    private var observer = this::syncView


    //just because we re-use these in 3 different button clicks
    //in this example we define them here
    private val success: Success = {
        Toast.makeText(
            context, "Success - you can use this trigger to " +
                    "perform a one off action like starting a new activity or " +
                    "something", Toast.LENGTH_SHORT
        ).show()
    }
    private val failureWithPayload: FailureWithPayload<UserMessage> = { userMessage ->
        Toast.makeText(
            context, "Fail - maybe tell the user to try again, message:" + userMessage.localisedMessage,
            Toast.LENGTH_SHORT
        ).show()
    }


    override fun onFinishInflate() {
        super.onFinishInflate()

        getModelReferences()

        setupButtonClickListeners()
    }

    private fun getModelReferences() {
        fruitFetcher = OG[FruitFetcher::class.java]
    }

    private fun setupButtonClickListeners() {
        fruit_fetchsuccess_btn.setOnClickListener { fruitFetcher.fetchFruits(success, failureWithPayload) }
        fruit_fetchchainsuccess_btn.setOnClickListener { fruitFetcher.fetchManyThings(success, failureWithPayload) }
        fruit_fetchfailbasic_btn.setOnClickListener { fruitFetcher.fetchFruitsButFail(success, failureWithPayload) }
        fruit_fetchfailadvanced_btn.setOnClickListener { fruitFetcher.fetchFruitsButFailAdvanced(success, failureWithPayload) }
    }


    //data binding stuff below

    fun syncView() {
        fruit_fetchsuccess_btn.isEnabled = !fruitFetcher.isBusy
        fruit_fetchchainsuccess_btn.isEnabled = !fruitFetcher.isBusy
        fruit_fetchfailbasic_btn.isEnabled = !fruitFetcher.isBusy
        fruit_fetchfailadvanced_btn.isEnabled = !fruitFetcher.isBusy
        fruit_name_textview.text = fruitFetcher.currentFruit.name
        fruit_citrus_img.setImageResource(
            if (fruitFetcher.currentFruit.isCitrus)
                R.drawable.lemon_positive else R.drawable.lemon_negative
        )
        fruit_tastyrating_tastybar.setTastyPercent(fruitFetcher.currentFruit.tastyPercentScore.toFloat())
        fruit_tastyrating_textview.text = String.format(
            context.getString(R.string.fruit_percent), fruitFetcher.currentFruit.tastyPercentScore.toString()
        )
        fruit_busy_progbar.visibility = if (fruitFetcher.isBusy)
            View.VISIBLE else View.INVISIBLE
        fruit_detailcontainer_linearlayout.visibility = if (fruitFetcher.isBusy)
            View.GONE else View.VISIBLE
    }


    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        fruitFetcher.addObserver(observer)
        syncView() //  <- don't forget this
    }


    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        fruitFetcher.removeObserver(observer)
    }
}
